### PR TITLE
Fwk 409 optimistic refresh initial token

### PR DESCRIPTION
**Pull request by Brady-Conn**
Ticket: https://greenchef.atlassian.net/browse/FWK-409

### CODE REVIEW

**Comment by bearfact** [comment]
I believe this can be simplified by using a setInterval that checks for the in-flight key, up to 10 times.  When the in-flight ke is gone it can assume the token is good, retrieve it, and go on its way.  I could be missing something

**Comment by Brady-Conn** [comment]
Oh yeah just checking for the in-flight key is a good idea, I'll try to refactor this a bit.

**Review by bearfact — CHANGES_REQUESTED** [review]
Let me know if you want to discuss those change requests.

**Comment by Brady-Conn** [comment]
Word, thanks! I think I figured out a pretty good solution with setInterval and checking for the refreshLockKey but I'd be happy to discuss further if this isn't close to what you had in mind. I also have it up on eph-brady-proxy.greenchef.com until the end of the day if you want to test anything.

**Comment by bearfact** [comment]
This is getting really hard to follow with all the nested conditional blocks.  Lets define what exactly this needs to do and refactor it to have some scoped functions/helpers.

Something like: 

- I have a valid token ( a token that exists and isn't about to expire)
  - send it on its way
- Else I don't have a valid token
  -  a reauth is in progress
    -- wait
  -   no reauth in progress
    -- get a new token

**Comment by Brady-Conn** [comment]
Sounds good, yeah you make a good point about all of the if statements.

**Comment by Brady-Conn** [comment]
I tried to reorganize this but I'm still not happy with the result. Maybe we could pair for a little tomorrow @bearfact to try and clean this up more?

**Comment by bearfact** [comment]
For sure.  Right after stand up.

**Comment by bearfact** [comment]
bump there version in policy and the package json and we're good

**Comment by bearfact** [comment]
:+1:

**Review by bearfact — APPROVED** [review]

**Review by tom-con — APPROVED** [review]

**Comment by tom-con** [comment]
👍
